Including lot chemist in creg meta data

hhan-schrodinger opened the pull request:
## Description
Users requested including lot chemist as an attribute for each record in creg metadata when calling the `cmpdReg/parentLot/getAllAuthorizedLots` endpoint with the goal of helping analyze and categorize records by using lot chemist. 


## How Has This Been Tested?
Tested locally by hit the route `http://localhost:3000/cmpdReg/parentLot/getAllAuthorizedLots`.

Before with test data, the route returned:
```
[
  {
    "id": 55,
    "lotCorpName": "CMPD-0000001-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000001",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000
  },
  {
    "id": 59,
    "lotCorpName": "CMPD-0000002-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000002",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000
  }
]
```

After the change, chemist was included in the response:
```
[
  {
    "chemist": "bob",
    "id": 55,
    "lotCorpName": "CMPD-0000001-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000001",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000
  },
  {
    "chemist": "bob",
    "id": 59,
    "lotCorpName": "CMPD-0000002-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000002",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000
  }
]
```

bffrost reviewed (changes requested):
@hhan-schrodinger in the comments on the ticket, Shawn also requested:
"In addition to lot chemist, the lot vendor and supplier would be useful for my work."

Can you add those too please?

hhan-schrodinger commented:
> @hhan-schrodinger in the comments on the ticket, Shawn also requested: "In addition to lot chemist, the lot vendor and supplier would be useful for my work."
> 
> Can you add those too please?

@bffrost - I tried adding vendor and supplier but ran into a strange issue where the route broke. I noticed vendor is a numeric column....since I'm not as familiar with the real world use case, would you suggest using the column vendor as is (a number), use "vendorid" instead (which is a string), or something else?

brianbolt commented on src/main/java/com/labsynch/labseer/domain/Lot.java:
```diff
@@ -614,7 +614,7 @@ public static TypedQuery<LotsByProjectDTO> findLotsByProjectsList(List<String> p
         EntityManager em = Lot.entityManager();
         String query = "SELECT new com.labsynch.labseer.dto.LotsByProjectDTO( "
-                + "lt.id as id, lt.corpName as lotCorpName, lt.lotNumber as lotNumber, lt.registrationDate as registrationDate, prnt.corpName as parentCorpName, lt.project as project) "
+                + "lt.id as id, lt.corpName as lotCorpName, lt.lotNumber as lotNumber, lt.registrationDate as registrationDate, prnt.corpName as parentCorpName, lt.project as project, lt.chemist as chemist) "
```
@hhan-schrodinger Per you question on the main thread. You should be able to join to vendor here.

e.g.
```
SELECT SELECT new com.labsynch.labseer.dto.LotsByProjectDTO(....etc. , vnd as vendor)
```

And:
```
+ "JOIN lt.vendor vnd "
```

Then you need to add vendor to the LotsByProjectDTO (initialization, getters and setters).

hhan-schrodinger replied:
@brianbolt I didn't realize it was a link to the vendor table, thanks!

hhan-schrodinger commented:
Fixed with supplier and vendor name, I added some test data in the database to make sure it worked. 

Example now when hitting the route:

```
[
  {
    "chemist": "bob",
    "id": 55,
    "lotCorpName": "CMPD-0000001-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000001",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000,
    "supplier": "supplier1",
    "vendor": "vendor1"
  },
  {
    "chemist": "bob",
    "id": 59,
    "lotCorpName": "CMPD-0000002-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000002",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000,
    "supplier": "supplier2",
    "vendor": "vendor2"
  }
]
```

brianbolt commented on src/main/java/com/labsynch/labseer/dto/LotsByProjectDTO.java:
```diff
@@ -13,21 +13,30 @@ public class LotsByProjectDTO {
-            String project) {
+            String project,
+            String chemist,
+            String vendor,
```
I think we should be specific that this is the vendorName not just vendor.  I also think we should include the vendor.code_name as "vendorCode"

@bffrost What do you think?

bffrost replied:
I agree with what @brianbolt wrote. Let's have `vendorName` and `vendorCode`. I think in most cases those are the same value but I can't imagine it'll hurt performance to have an extra field from that same table, so seems better to have more than to have less

hhan-schrodinger replied:
Updated with the changes, tested the route and it looks good:

```
[
  {
    "chemist": "bob",
    "id": 55,
    "lotCorpName": "CMPD-0000001-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000001",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000,
    "supplier": "supplier1",
    "vendorCode": "somecode",
    "vendorName": "vendor1"
  },
  {
    "chemist": "bob",
    "id": 59,
    "lotCorpName": "CMPD-0000002-001",
    "lotNumber": 1,
    "parentCorpName": "CMPD-0000002",
    "project": "PROJ-00000001",
    "registrationDate": 1675314000000,
    "supplier": "supplier2",
    "vendorCode": "somecode2",
    "vendorName": "vendor2"
  }
]
```

bffrost commented on src/main/java/com/labsynch/labseer/domain/Lot.java:
```diff
@@ -614,10 +614,11 @@ public static TypedQuery<LotsByProjectDTO> findLotsByProjectsList(List<String> p
                 + "FROM Lot AS lt "
                 + "JOIN lt.saltForm sltfrm "
                 + "JOIN sltfrm.parent prnt "
+                + "JOIN lt.vendor vnd "
```
I think we want to do a `LEFT JOIN` here so we don't filter out lots with no vendor

hhan-schrodinger replied:
Good catch, did not think of that - thanks!

hhan-schrodinger replied:
Fixed it, tested it by removing vendor for a lot then hitting the API route, looks good!

bffrost reviewed (changes requested):
One important change needed for the join, but otherwise LGTM.

@hhan-schrodinger it would be a good idea for you to reach out to Shawn Watts to tell him you're working on this and just confirm there aren't any other lot fields that'd be useful to him. Just because the delay between this dev work and when it gets deployed is so long, it'd be good not to "oops" miss any fields that we could've added at this point.

hhan-schrodinger commented:
@bffrost Sounds good, I've reached out to Shawn Watts, will keep this PR open for a bit in case he asks for any other fields.

hhan-schrodinger commented:
Closing this after being informed by Shawn that he doesn't need more fields.

bffrost reviewed (approved):
LGTM!
